contrib/go/src/python/pants/contrib/go/tasks/go_protobuf_gen.py

```diff
@@ -0,0 +1,127 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+import re
+
+from pants.backend.codegen.protobuf.subsystems.protoc import Protoc
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.build_graph.address import Address
+from pants.option.custom_types import target_option
+from pants.task.simple_codegen_task import SimpleCodegenTask
+from pants.util.dirutil import safe_mkdir
+from pants.util.memo import memoized_property
+from pants.util.process_handler import subprocess
+from twitter.common.collections import OrderedSet
+
+from pants.contrib.go.subsystems.protoc_gen_go import ProtocGenGo
+from pants.contrib.go.targets.go_protobuf_library import GoProtobufGenLibrary, GoProtobufLibrary
+
+
+class GoProtobufGen(SimpleCodegenTask):
+
+  _NAMESPACE_PARSER = re.compile(r'^\s*option\s+go_package\s*=\s*"([^\s]+)"\s*;', re.MULTILINE)
+
+  @classmethod
+  def register_options(cls, register):
+    super(GoProtobufGen, cls).register_options(register)
+
+    register('--import-target', type=target_option, advanced=True,
+             help='Target that will be added as a dependency of generated code.')
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(GoProtobufGen, cls).subsystem_dependencies() + (Protoc.scoped(cls), ProtocGenGo.scoped(cls),)
+
+  @property
+  def _binary(self):
+    return self._protoc.select(context=self.context)
+
+  @property
+  def _version(self):
+    return self._protoc.version(context=self.context)
+
+  @memoized_property
+  def _protoc(self):
+    return Protoc.scoped_instance(self)
+
+  @memoized_property
+  def _protoc_gen_go(self):
+    return ProtocGenGo.scoped_instance(self)
+
+  def synthetic_target_extra_dependencies(self, target, target_workdir):
+    import_target = self.get_options().import_target
+    if import_target is None:
+      raise TaskError('Option import_target in scope {} must be set.'.format(
+        self.options_scope))
+    return self.context.resolve(import_target)
+
+  def synthetic_target_type(self, target):
+    return GoProtobufGenLibrary
+
+  def is_gentarget(self, target):
+    return isinstance(target, GoProtobufLibrary)
+
+  @classmethod
+  def product_types(cls):
+    return ['go']
+
+  def execute_codegen(self, target, target_workdir):
+    target_cmd = [self._binary]
+
+    env = os.environ.copy()
+    env['PATH'] = ':'.join([env['PATH'],
+                            os.path.dirname(self._protoc_gen_go.select(context=self.context))])
+
+    bases = OrderedSet(tgt.target_base for tgt in target.closure() if self.is_gentarget(tgt))
+    for base in bases:
+      target_cmd.append('-I={}'.format(os.path.join(get_buildroot(), base)))
```

Comment:
So, the java task does this as well (giving the entire target base to `protoc`), but if there is any way to hand it explicit files for the dependencies, that would be ideal (it might not be). Giving the compiler the entire base can lead to undeclared dependencies.